File: problems/src/clojure/src/ch1_arrays_and_strings/ch1_5.clj
(ns ch1-arrays-and-strings.ch1-5)

; fairly direct translation of Java algorithm
; not very pleasant to look at and it seems messy
(defn- compress-string [input]
  (let [output (loop [current (first input)
                      currentCount 0
                      in (seq input)
                      out []]
                 ;(prn current currentCount in out)
                 (cond (and (= current (first in))          ; same as the pointer and not the last element
                            (not= 1 (count in)))
                       (recur current (inc currentCount) (next in) out) ; just increment count and move on
                       (and (= current (first in))          ; same as pointer and the last element
                            (= 1 (count in)))
                       (into out [current (inc currentCount)])

                       (and (not= current (first in))       ; different from pointer and not the last element
                            (not= 1 (count in)))
                       (recur (first in) 1 (rest in) (into out [current currentCount]))
                       (and (not= current (first in))       ; different from pointer and the last element
                            (= 1 (count in)))
                       (into out [current currentCount (first in) 1])
                       ))]
    (if (>= (count output) (count input))
      input
      (apply str output))))

(defn compress-old [input]
  (if (or (nil? input)
          (<= (count input) 2))
    input
    (let [compressed (compress-string-2 input)]
      (if (>= (count compressed) (count input))
        input
        (apply str compressed)))))

(ffirst )

(defn- compress-string-2 [current
                          currentCount
                          in
                          out]
  (cond
    ; If the input is empty, just add the current tracker and count to the output and return it
    (empty? in)
    (into out [current currentCount])
    ; If the input is the same as the tracker, increment the counter and continue
    (= current (first in))
    (recur current (inc currentCount) (rest in) out)
    ; If the input is different from the tracker, add the current values to the output
    ; and reset the tracker and count
    (not= current (first in))
    (recur (first in) 1 (rest in) (into out [current currentCount]))))

(defn compress [input]
  (if (or (nil? input)
          (<= (count input) 2))
    ; Check to see if the input is worth compressing
    input
    (let [compressed (compress-string-2 (first input) 0 (seq input) [])
          ; Check to see if the output was an improvement
          use-compressed? (>= (count compressed) (count input))]
      (if use-compressed?
        input
        (apply str compressed)))))

(compress "aabbbcc")
(compress "abc")
(compress "aaa")
(compress "aaabbbccc")
(compress "aaaabababacc")
(compress "aaaabababaccccccccccccc")
(compress "aaaabababacccccccccccccd")

File: problems/src/java/src/ch1_arrays_and_strings/Ch1_5.java
package ch1_arrays_and_strings;

public class Ch1_5 {

    public static void main(String[] args) {
        System.out.println(compress("aabbbcc"));
        System.out.println(compress("abc"));
        System.out.println(compress("aaa"));
        System.out.println(compress("aaaabababacc"));
        System.out.println(compress("aaaabababaccccccccccccc"));
        System.out.println(compress("aaaabababacccccccccccccd"));
    }

    public static String compress(String input) {
        if (input == null || input.length() <= 2) {
            // invalid or too short for compression to be of value
            return input;
        }
        StringBuilder sb = new StringBuilder();
        char[] in = input.toCharArray();

        char current = in[0];
        int currentCount = 1;
        for (int i = 0; i < in.length; i++) {
            if (current == in[i] && i != in.length - 1) {
                // is the same and is not the very last
                currentCount++;
            } else if (current == in[i] && i == in.length - 1) {
                // is the same and is the very last
                currentCount++;
                sb.append(current);
                sb.append(currentCount);
            } else if (i == in.length - 1 && current != in[i]) {
                // input value is different and its the end
                sb.append(current);
                sb.append(currentCount);
                current = in[i];
                currentCount = 1;
                sb.append(current);
                sb.append(currentCount);
            } else {
                sb.append(current);
                sb.append(currentCount);
                current = in[i];
                currentCount = 1;
            }
        }

        if (sb.length() >= input.length()) {
            return input;
        }
        return sb.toString();
    }
}
